coredb-operator/src/statefulset.rs

```diff
@@ -306,79 +307,133 @@ pub fn stateful_set_from_cdb(cdb: &CoreDB) -> StatefulSet {
+                pvcs_to_update.push((share_pvc_name, cdb.spec.sharedirStorage.clone()));
             }
             if cdb.status.clone().unwrap().pkglibdirStorage != cdb.spec.pkglibdirStorage {
-                pvcs_to_update.push(("pkglibdir".to_string(), cdb.spec.pkglibdirStorage.clone()));
+                pvcs_to_update.push((pkglib_pvc_name, cdb.spec.pkglibdirStorage.clone()));
             }
-            pvcs_to_update
+            debug!("pvcs_to_update: {:?}", pvcs_to_update);
+            (pvcs_to_update, pvcs_to_create)
+        }
+        false => {
+            debug!("cdb is not running");
+            (vec![], vec![])
         }
-        false => Vec::new(),
     };
-    debug!("pvcs_to_update: {:?}", pvcs_to_update);
-    if !pvcs_to_update.is_empty() {
-        let sts_name = sts.clone().metadata.name.unwrap();
-        let sts_namespace = sts.clone().metadata.namespace.unwrap();
 
-        // why do delete first, then update?
-        delete_sts_no_cascade(&sts_api, &sts_name).await;
 
+    if !pvcs_to_update.is_empty() {
+        delete_sts_no_cascade(&sts_api, &sts_name).await?;
         let pvc_api: Api<PersistentVolumeClaim> = Api::namespaced(ctx.client.clone(), &sts_namespace);
-        for (pvc_name, qty) in pvcs_to_update {
-            update_pvc(&pvc_api, &sts_name, &pvc_name, qty).await;
+        for (pvc_full_name, qty) in pvcs_to_update {
+            update_pvc(&pvc_api, &pvc_full_name, qty).await;
         }
     }
 
+    if !pvcs_to_create.is_empty() {
+        delete_sts_no_cascade(&sts_api, &sts_name).await?;
+        let primary_pod = cdb.primary_pod(client.clone()).await?;
+        let pod_api: Api<Pod> = Api::namespaced(client.clone(), &sts_namespace);
+
+        let prim_pod_name = primary_pod.metadata.name.unwrap();
+        warn!("deleting pod to attach pvc: {}", prim_pod_name);
+        pod_api.delete(&prim_pod_name, &DeleteParams::default()).await?;
+    }
+
     let ps = PatchParams::apply("cntrlr").force();
     let _o = sts_api
         .patch(&sts.clone().metadata.name.unwrap(), &ps, &Patch::Apply(&sts))
         .await
         .map_err(Error::KubeError)?;
-
     Ok(())
 }
 
-async fn delete_sts_no_cascade(sts_api: &Api<StatefulSet>, sts_name: &str) {
+async fn delete_sts_no_cascade(sts_api: &Api<StatefulSet>, sts_name: &str) -> Result<(), Error> {
     let delete_params: DeleteParams = DeleteParams {
         dry_run: false,
         grace_period_seconds: None,
         propagation_policy: Some(kube::api::PropagationPolicy::Orphan),
         preconditions: None,
     };
-
-    let _o = sts_api
+    info!("deleting_sts_no_cascade: {}", sts_name);
+    let _ = sts_api
         .delete(sts_name, &delete_params)
         .await
-        .map_err(Error::KubeError);
+        .map_err(Error::KubeError)?;
+    thread::sleep(Duration::from_millis(3000));
+    Ok(())
+}
+
+fn pvc_full_name(pvc_name: &str, sts_name: &str) -> String {
+    format!("{pvc_name}-{sts_name}-0")
 }
```

Comment:
moved this logic out, so we can create and use the pvc names upstream in the pvc reconcile process